File: adapters/python/cup.py
"""CUP Python adapter."""

from __future__ import annotations

import json
import re
from dataclasses import dataclass, field
from typing import Any, Literal

ADAPTER_VERSION = "cup-python/0.3.0"
PROVENANCE_EXTENSION = "cup.provenance"
SUPPORTED_METHODS = {"GET", "POST", "PUT", "PATCH", "DELETE"}
SCRIPT_TAG_PATTERN = re.compile(r"<script\b", re.IGNORECASE)
INLINE_HANDLER_PATTERN = re.compile(r"\son[a-z][a-z0-9_-]*\s*=", re.IGNORECASE)
JAVASCRIPT_URL_PATTERN = re.compile(r"\b(?:href|src)\s*=\s*(['\"])\s*javascript:", re.IGNORECASE)
SAFE_FILTER_PATTERN = re.compile(r"\|\s*safe\b")
RELATIVE_URL_PATTERN = re.compile(r"^(\/(?!\/)|\.{1,2}\/|[?#])")


# ── Action descriptors ────────────────────────────────────────────────────────

@dataclass
class FetchAction:
    url: str
    method: Literal["GET", "POST", "PUT", "PATCH", "DELETE"] = "POST"
    payload: dict[str, Any] = field(default_factory=dict)
    type: str = field(init=False, default="fetch")

    def to_dict(self) -> dict[str, Any]:
        d: dict[str, Any] = {"type": self.type, "url": self.url, "method": self.method}
        if self.payload:
            d["payload"] = self.payload
        return d


@dataclass
class EmitAction:
    event: str
    detail: dict[str, Any] = field(default_factory=dict)
    type: str = field(init=False, default="emit")

    def to_dict(self) -> dict[str, Any]:
        d: dict[str, Any] = {"type": self.type, "event": self.event}
        if self.detail:
            d["detail"] = self.detail
        return d


@dataclass
class NavigateAction:
    url: str
    replace: bool = False
    type: str = field(init=False, default="navigate")

    def to_dict(self) -> dict[str, Any]:
        return {"type": self.type, "url": self.url, "replace": self.replace}


Action = FetchAction | EmitAction | NavigateAction


class ValidationError(ValueError):
    def __init__(self, issues: list[str]) -> None:
        self.issues = issues
        super().__init__("invalid CUP protocol view: " + "; ".join(issues))


@dataclass(frozen=True)
class ViewPolicy:
    require_version: bool = False
    require_title: bool = False
    require_route: bool = False
    allow_safe_filter: bool = True
    allow_inline_handlers: bool = True
    allow_javascript_urls: bool = True
    allow_script_tags: bool = True
    action_urls: Literal["relative-only", "any"] = "any"


STARTER_VIEW_POLICY = ViewPolicy(
    require_version=True,
    require_title=True,
    require_route=True,
    allow_safe_filter=False,
    allow_inline_handlers=False,
    allow_javascript_urls=False,
    allow_script_tags=False,
    action_urls="relative-only",
)


class PolicyError(ValueError):
    def __init__(self, issues: list[str]) -> None:
        self.issues = issues
        super().__init__("CUP view policy rejected: " + "; ".join(issues))


# ── UIView builder ────────────────────────────────────────────────────────────

class UIView:
    """
    Fluent builder for the CUP UIView contract.

    All methods return `self` for chaining.
    Call `.to_dict()` or `.to_response()` at the end.
    """

    def __init__(self, template: str) -> None:
        self._template = template
        self._state: dict[str, Any] = {}
        self._actions: dict[str, Action] = {}
        self._title: str | None = None
        self._route: str | None = None

    # ── Builder methods ───────────────────────────────────────────────────────

    def state(self, **kwargs: Any) -> "UIView":
        """Merge keyword arguments into the view state."""
        self._state.update(kwargs)
        return self

    def state_dict(self, data: dict[str, Any]) -> "UIView":
        """Merge a plain dict into the view state."""
        self._state.update(data)
        return self

    def action(self, name: str, descriptor: Action) -> "UIView":
        """Register a named action."""
        self._actions[name] = descriptor
        return self

    def title(self, t: str) -> "UIView":
        self._title = t
        return self

    def route(self, r: str) -> "UIView":
        self._route = r
        return self

    # ── Serialisation ─────────────────────────────────────────────────────────

    def to_dict(self) -> dict[str, Any]:
        """Return the UIView as a plain Python dict."""
        out: dict[str, Any] = {
            "template": self._template,
            "state": self._state,
        }
        if self._actions:
            out["actions"] = {k: v.to_dict() for k, v in self._actions.items()}

        meta: dict[str, Any] = {"version": "1", "lang": "python", "generator": ADAPTER_VERSION}
        if self._title:
            meta["title"] = self._title
        if self._route:
            meta["route"] = self._route
        meta["provenance"] = {
            "source": "adapter",
            "generatedBy": ADAPTER_VERSION,
            "generatedAt": _iso_now(),
            "validation": {
                "schema": "valid",
                "policy": "skipped",
                "validator": ADAPTER_VERSION,
                "checkedAt": _iso_now(),
            },
        }
        meta["extensions"] = {
            PROVENANCE_EXTENSION: {"version": "1"},
        }
        out["meta"] = meta

        validate_view(out)
        return out

    def to_json(self, **kwargs: Any) -> str:
        """Serialise to a JSON string."""
        return json.dumps(self.to_dict(), **kwargs)

    def to_response(self) -> tuple[str, str]:
        """
        Return (json_body, content_type) ready to hand to any HTTP framework.

        Example — stdlib:
            body, ct = view.to_response()
            self.send_response(200)
            self.send_header("Content-Type", ct)
            self.end_headers()
            self.wfile.write(body.encode())

        Example — Flask:
            return Response(*view.to_response())

        Example — Django:
            body, ct = view.to_response()
            return HttpResponse(body, content_type=ct)

        Example — FastAPI / Starlette:
            body, ct = view.to_response()
            return Response(content=body, media_type=ct)
        """
        return self.to_json(), "application/json"


def validate_view(view: UIView | dict[str, Any]) -> dict[str, Any]:
    if isinstance(view, UIView):
        payload = {
            "template": view._template,
            "state": view._state,
            "actions": {k: v.to_dict() for k, v in view._actions.items()} if view._actions else None,
            "meta": {
                "version": "1",
                "lang": "python",
                "generator": ADAPTER_VERSION,
                **({"title": view._title} if view._title else {}),
                **({"route": view._route} if view._route else {}),
                "provenance": {
                    "source": "adapter",
                    "generatedBy": ADAPTER_VERSION,
                    "generatedAt": _iso_now(),
                    "validation": {
                        "schema": "valid",
                        "policy": "skipped",
                        "validator": ADAPTER_VERSION,
                        "checkedAt": _iso_now(),
                    },
                },
                "extensions": {
                    PROVENANCE_EXTENSION: {"version": "1"},
                },
            },
        }
        if not payload["actions"]:
            payload.pop("actions")
    else:
        payload = view

    issues: list[str] = []
    _validate_payload(_normalise(payload), "view", issues)
    if issues:
        raise ValidationError(issues)
    _annotate_schema_provenance(payload)
    return payload


def validate_view_policy(view: UIView | dict[str, Any], policy: ViewPolicy | None = None) -> dict[str, Any]:
    payload = validate_view(view)
    effective = policy or ViewPolicy()
    issues: list[str] = []
    _validate_policy_payload(payload, "view", effective, issues)
    if issues:
        raise PolicyError(issues)
    _annotate_policy_provenance(payload, "starter-view-policy" if policy is STARTER_VIEW_POLICY else "custom-view-policy")
    return payload


def _normalise(value: Any) -> Any:
    try:
        return json.loads(json.dumps(value))
    except TypeError as exc:
        raise ValidationError([f"value is not JSON-serializable: {exc}"]) from exc


def _validate_payload(value: Any, path: str, issues: list[str]) -> None:
    if not isinstance(value, dict):
        issues.append(f"{path} must be an object")
        return

    _validate_allowed_keys(value, {"template", "state", "actions", "meta"}, path, issues)

    if not isinstance(value.get("template"), str):
        issues.append(f"{path}.template must be a string")

    state = value.get("state")
    if not isinstance(state, dict):
        issues.append(f"{path}.state must be an object")
    else:
        _validate_json_object(state, f"{path}.state", issues)

    actions = value.get("actions")
    if actions is not None:
        if not isinstance(actions, dict):
            issues.append(f"{path}.actions must be an object")
        else:
            for name, descriptor in actions.items():
                _validate_action(descriptor, f"{path}.actions.{name}", issues)

    meta = value.get("meta")
    if meta is not None:
        _validate_meta(meta, f"{path}.meta", issues)


def _validate_action(value: Any, path: str, issues: list[str]) -> None:
    if not isinstance(value, dict):
        issues.append(f"{path} must be an object")
        return

    kind = value.get("type")
    if kind == "fetch":
        _validate_allowed_keys(value, {"type", "url", "method", "payload"}, path, issues)
        if not isinstance(value.get("url"), str):
            issues.append(f"{path}.url must be a string")
        method = value.get("method")
        if method is not None and method not in SUPPORTED_METHODS:
            issues.append(f"{path}.method must be a supported HTTP method")
        payload = value.get("payload")
        if payload is not None:
            if not isinstance(payload, dict):
                issues.append(f"{path}.payload must be an object")
            else:
                _validate_json_object(payload, f"{path}.payload", issues)
        return

    if kind == "emit":
        _validate_allowed_keys(value, {"type", "event", "detail"}, path, issues)
        if not isinstance(value.get("event"), str):
            issues.append(f"{path}.event must be a string")
        detail = value.get("detail")
        if detail is not None:
            if not isinstance(detail, dict):
                issues.append(f"{path}.detail must be an object")
            else:
                _validate_json_object(detail, f"{path}.detail", issues)
        return

    if kind == "navigate":
        _validate_allowed_keys(value, {"type", "url", "replace"}, path, issues)
        if not isinstance(value.get("url"), str):
            issues.append(f"{path}.url must be a string")
        replace = value.get("replace")
        if replace is not None and not isinstance(replace, bool):
            issues.append(f"{path}.replace must be a boolean")
        return

    issues.append(f"{path}.type must be one of fetch, emit, navigate")


def _validate_meta(value: Any, path: str, issues: list[str]) -> None:
    if not isinstance(value, dict):
        issues.append(f"{path} must be an object")
        return

    _validate_allowed_keys(value, {"version", "lang", "generator", "title", "route", "provenance", "extensions"}, path, issues)

    if "version" in value and value["version"] != "1":
        issues.append(f"{path}.version must be '1'")

    for key in ("lang", "generator", "title", "route"):
        if key in value and not isinstance(value[key], str):
            issues.append(f"{path}.{key} must be a string")

    provenance = value.get("provenance")
    if provenance is not None:
        _validate_provenance(provenance, f"{path}.provenance", issues)

    extensions = value.get("extensions")
    if extensions is not None:
        if not isinstance(extensions, dict):
            issues.append(f"{path}.extensions must be an object")
        else:
            for name, descriptor in extensions.items():
                _validate_extension_descriptor(descriptor, f"{path}.extensions.{name}", issues)


def _validate_json_object(value: dict[str, Any], path: str, issues: list[str]) -> None:
    for key, item in value.items():
        _validate_json_value(item, f"{path}.{key}", issues)


def _validate_json_value(value: Any, path: str, issues: list[str]) -> None:
    if value is None or isinstance(value, (str, int, float, bool)):
        return
    if isinstance(value, list):
        for index, item in enumerate(value):
            _validate_json_value(item, f"{path}[{index}]", issues)
        return
    if isinstance(value, dict):
        for key, item in value.items():
            _validate_json_value(item, f"{path}.{key}", issues)
        return
    issues.append(f"{path} must be JSON-serializable")


def _validate_allowed_keys(value: dict[str, Any], allowed: set[str], path: str, issues: list[str]) -> None:
    for key in value:
        if key not in allowed:
            issues.append(f"{path} contains unsupported property '{key}'")


def _validate_provenance(value: Any, path: str, issues: list[str]) -> None:
    if not isinstance(value, dict):
        issues.append(f"{path} must be an object")
        return

    _validate_allowed_keys(value, {"source", "generatedBy", "generatedAt", "requestId", "validation", "policyDecisions"}, path, issues)

    if "source" in value and value["source"] not in {"human", "ai", "adapter", "hybrid"}:
        issues.append(f"{path}.source must be one of human, ai, adapter, hybrid")
    for key in ("generatedBy", "generatedAt", "requestId"):
        if key in value and not isinstance(value[key], str):
            issues.append(f"{path}.{key} must be a string")

    validation = value.get("validation")
    if validation is not None:
        _validate_validation_provenance(validation, f"{path}.validation", issues)

    policy_decisions = value.get("policyDecisions")
    if policy_decisions is not None:
        if not isinstance(policy_decisions, list):
            issues.append(f"{path}.policyDecisions must be an array")
        else:
            for index, decision in enumerate(policy_decisions):
                _validate_policy_decision(decision, f"{path}.policyDecisions[{index}]", issues)


def _validate_validation_provenance(value: Any, path: str, issues: list[str]) -> None:
    if not isinstance(value, dict):
        issues.append(f"{path} must be an object")
        return

    _validate_allowed_keys(value, {"schema", "policy", "validator", "checkedAt"}, path, issues)

    if value.get("schema") not in {"valid", "repaired", "unchecked"}:
        issues.append(f"{path}.schema must be one of valid, repaired, unchecked")
    if "policy" in value and value["policy"] not in {"passed", "failed", "skipped"}:
        issues.append(f"{path}.policy must be one of passed, failed, skipped")
    for key in ("validator", "checkedAt"):
        if key in value and not isinstance(value[key], str):
            issues.append(f"{path}.{key} must be a string")


def _validate_policy_decision(value: Any, path: str, issues: list[str]) -> None:
    if not isinstance(value, dict):
        issues.append(f"{path} must be an object")
        return

    _validate_allowed_keys(value, {"policy", "outcome", "detail"}, path, issues)

    if not isinstance(value.get("policy"), str):
        issues.append(f"{path}.policy must be a string")
    if value.get("outcome") not in {"allow", "deny", "skip"}:
        issues.append(f"{path}.outcome must be one of allow, deny, skip")
    if "detail" in value and not isinstance(value["detail"], str):
        issues.append(f"{path}.detail must be a string")


def _validate_extension_descriptor(value: Any, path: str, issues: list[str]) -> None:
    if not isinstance(value, dict):
        issues.append(f"{path} must be an object")
        return

    _validate_allowed_keys(value, {"version", "required", "config"}, path, issues)

    if not isinstance(value.get("version"), str):
        issues.append(f"{path}.version must be a string")
    if "required" in value and not isinstance(value["required"], bool):
        issues.append(f"{path}.required must be a boolean")
    config = value.get("config")
    if config is not None:
        if not isinstance(config, dict):
            issues.append(f"{path}.config must be an object")
        else:
            _validate_json_object(config, f"{path}.config", issues)


def _validate_policy_payload(value: dict[str, Any], path: str, policy: ViewPolicy, issues: list[str]) -> None:
    meta = value.get("meta") if isinstance(value.get("meta"), dict) else {}
    template = value.get("template", "")
    actions = value.get("actions") if isinstance(value.get("actions"), dict) else {}

    if policy.require_version and meta.get("version") != "1":
        issues.append(f"{path}.meta.version is required by policy")
    if policy.require_title and not meta.get("title"):
        issues.append(f"{path}.meta.title is required by policy")
    if policy.require_route and not meta.get("route"):
        issues.append(f"{path}.meta.route is required by policy")
    if not policy.allow_safe_filter and isinstance(template, str) and SAFE_FILTER_PATTERN.search(template):
        issues.append(f"{path}.template uses the |safe filter, which is disabled by policy")
    if not policy.allow_script_tags and isinstance(template, str) and SCRIPT_TAG_PATTERN.search(template):
        issues.append(f"{path}.template contains a <script> tag, which is disabled by policy")
    if not policy.allow_inline_handlers and isinstance(template, str) and INLINE_HANDLER_PATTERN.search(template):
        issues.append(f"{path}.template contains inline event handler attributes, which are disabled by policy")
    if not policy.allow_javascript_urls and isinstance(template, str) and JAVASCRIPT_URL_PATTERN.search(template):
        issues.append(f"{path}.template contains a javascript: URL, which is disabled by policy")

    if policy.action_urls == "relative-only":
        for name, descriptor in actions.items():
            _validate_policy_action(descriptor, f"{path}.actions.{name}", issues)


def _validate_policy_action(value: Any, path: str, issues: list[str]) -> None:
    if not isinstance(value, dict):
        return
    if value.get("type") not in {"fetch", "navigate"}:
        return
    url = value.get("url")
    if not isinstance(url, str) or not RELATIVE_URL_PATTERN.match(url):
        issues.append(f"{path}.url must stay relative under the current policy")


def _annotate_schema_provenance(payload: dict[str, Any]) -> None:
    meta = payload.setdefault("meta", {})
    if not isinstance(meta, dict):
        return

    provenance = meta.setdefault("provenance", {})
    if not isinstance(provenance, dict):
        return

    provenance.setdefault("source", "adapter")
    provenance.setdefault("generatedBy", meta.get("generator"))
    provenance.setdefault("generatedAt", _iso_now())
    validation = provenance.setdefault("validation", {})
    if isinstance(validation, dict):
        validation["schema"] = "valid"
        validation["policy"] = validation.get("policy", "skipped")
        validation["validator"] = validation.get("validator", meta.get("generator", ADAPTER_VERSION))
        validation["checkedAt"] = _iso_now()

    extensions = meta.setdefault("extensions", {})
    if isinstance(extensions, dict):
        extensions.setdefault(PROVENANCE_EXTENSION, {"version": "1"})


def _annotate_policy_provenance(payload: dict[str, Any], policy_name: str) -> None:
    meta = payload.get("meta")
    if not isinstance(meta, dict):
        return

    provenance = meta.setdefault("provenance", {})
    if not isinstance(provenance, dict):
        return

    validation = provenance.setdefault("validation", {})
    if isinstance(validation, dict):
        validation["policy"] = "passed"
        validation["checkedAt"] = _iso_now()

    decisions = provenance.setdefault("policyDecisions", [])
    if isinstance(decisions, list):
        decisions.append({"policy": policy_name, "outcome": "allow"})


def _iso_now() -> str:
    from datetime import datetime, timezone

    return datetime.now(timezone.utc).isoformat()

File: adapters/go/cup.go
// Package cup provides the Go adapter for the CUP UI Protocol.
// Any net/http handler can call view.WriteJSON(w) to serve a UIView.
//
// Usage:
//
//	view := cup.New(`<h1>{{ title }}</h1><p>{{ body }}</p>`).
//	    State(cup.S{"title": "Hello", "body": "From Go"}).
//	    Action("reload", cup.Fetch("/api/reload")).
//	    Title("Home").
//	    Route("/")
//
//	http.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {
//	    view.WriteJSON(w)
//	})
package cup

import (
	"encoding/json"
	"fmt"
	"net/http"
	"regexp"
	"strings"
)

const adapterVersion = "cup-go/0.2.4"

var (
	scriptTagPattern     = regexp.MustCompile(`(?i)<script\b`)
	inlineHandlerPattern = regexp.MustCompile(`(?i)\son[a-z][a-z0-9_-]*\s*=`)
	javascriptURLPattern = regexp.MustCompile(`(?i)\b(?:href|src)\s*=\s*(['"])\s*javascript:`)
	safeFilterPattern    = regexp.MustCompile(`\|\s*safe\b`)
)

// S is a convenience alias for map[string]any, used for state values.
type S = map[string]any

// ── Action descriptors ────────────────────────────────────────────────────────

type actionKind string

const (
	kindFetch    actionKind = "fetch"
	kindEmit     actionKind = "emit"
	kindNavigate actionKind = "navigate"
)

// ActionDescriptor is the JSON representation of any action.
type ActionDescriptor struct {
	Type    actionKind `json:"type"`
	URL     string     `json:"url,omitempty"`
	Method  string     `json:"method,omitempty"`
	Payload S          `json:"payload,omitempty"`
	Event   string     `json:"event,omitempty"`
	Detail  S          `json:"detail,omitempty"`
	Replace bool       `json:"replace,omitempty"`
}

// Fetch creates a FetchAction that POSTs to the given URL.
func Fetch(url string, opts ...func(*ActionDescriptor)) ActionDescriptor {
	a := ActionDescriptor{Type: kindFetch, URL: url, Method: "POST"}
	for _, o := range opts {
		o(&a)
	}
	return a
}

// WithMethod overrides the HTTP method on a FetchAction.
func WithMethod(m string) func(*ActionDescriptor) {
	return func(a *ActionDescriptor) { a.Method = m }
}

// WithPayload adds extra keys to the request body of a FetchAction.
func WithPayload(p S) func(*ActionDescriptor) {
	return func(a *ActionDescriptor) { a.Payload = p }
}

// Emit creates an EmitAction that dispatches a CustomEvent on the container.
func Emit(event string, detail ...S) ActionDescriptor {
	a := ActionDescriptor{Type: kindEmit, Event: event}
	if len(detail) > 0 {
		a.Detail = detail[0]
	}
	return a
}

// Navigate creates a NavigateAction that uses the History API.
func Navigate(url string, replace ...bool) ActionDescriptor {
	a := ActionDescriptor{Type: kindNavigate, URL: url}
	if len(replace) > 0 {
		a.Replace = replace[0]
	}
	return a
}

// ── UIView ────────────────────────────────────────────────────────────────────

type meta struct {
	Version   string `json:"version"`
	Lang      string `json:"lang"`
	Generator string `json:"generator"`
	Title     string `json:"title,omitempty"`
	Route     string `json:"route,omitempty"`
}

type uiViewJSON struct {
	Template string                      `json:"template"`
	State    S                           `json:"state"`
	Actions  map[string]ActionDescriptor `json:"actions,omitempty"`
	Meta     meta                        `json:"meta"`
}

type ValidationError struct {
	Issues []string
}

func (e ValidationError) Error() string {
	return "invalid CUP protocol view: " + strings.Join(e.Issues, "; ")
}

type ActionURLPolicy string

const (
	ActionURLsAny          ActionURLPolicy = "any"
	ActionURLsRelativeOnly ActionURLPolicy = "relative-only"
)

type ViewPolicy struct {
	RequireVersion      bool
	RequireTitle        bool
	RequireRoute        bool
	AllowSafeFilter     bool
	AllowInlineHandlers bool
	AllowJavaScriptURLs bool
	AllowScriptTags     bool
	ActionURLs          ActionURLPolicy
}

var StarterViewPolicy = ViewPolicy{
	RequireVersion:      true,
	RequireTitle:        true,
	RequireRoute:        true,
	AllowSafeFilter:     false,
	AllowInlineHandlers: false,
	AllowJavaScriptURLs: false,
	AllowScriptTags:     false,
	ActionURLs:          ActionURLsRelativeOnly,
}

type PolicyError struct {
	Issues []string
}

func (e PolicyError) Error() string {
	return "CUP view policy rejected: " + strings.Join(e.Issues, "; ")
}

// UIView is the CUP UI contract builder.
type UIView struct {
	template string
	state    S
	actions  map[string]ActionDescriptor
	title    string
	route    string
}

// New creates a new UIView with the given Django-style template.
func New(template string) *UIView {
	return &UIView{
		template: template,
		state:    S{},
		actions:  map[string]ActionDescriptor{},
	}
}

// State merges key/value pairs into the view state.
func (v *UIView) State(data S) *UIView {
	for k, val := range data {
		v.state[k] = val
	}
	return v
}

// Action registers a named action.
func (v *UIView) Action(name string, a ActionDescriptor) *UIView {
	v.actions[name] = a
	return v
}

// Title sets the view title (shown in CUP DevTools).
func (v *UIView) Title(t string) *UIView {
	v.title = t
	return v
}

// Route records the server route that produced this view.
func (v *UIView) Route(r string) *UIView {
	v.route = r
	return v
}

// ToMap returns the UIView as a plain map ready for JSON marshalling.
func (v *UIView) ToMap() uiViewJSON {
	return uiViewJSON{
		Template: v.template,
		State:    v.state,
		Actions:  v.actions,
		Meta: meta{
			Version:   "1",
			Lang:      "go",
			Generator: adapterVersion,
			Title:     v.title,
			Route:     v.route,
		},
	}
}

func (v *UIView) Validate() error {
	return Validate(v.ToMap())
}

// MarshalJSON implements json.Marshaler.
func (v *UIView) MarshalJSON() ([]byte, error) {
	if err := v.Validate(); err != nil {
		return nil, err
	}
	return json.Marshal(v.ToMap())
}

// WriteJSON writes the UIView as JSON to an http.ResponseWriter.
// Sets Content-Type to application/json and status 200.
func (v *UIView) WriteJSON(w http.ResponseWriter) error {
	if err := v.Validate(); err != nil {
		return err
	}
	b, err := json.Marshal(v)
	if err != nil {
		return err
	}
	w.Header().Set("Content-Type", "application/json")
	_, err = w.Write(b)
	return err
}

func Validate(input any) error {
	encoded, err := json.Marshal(input)
	if err != nil {
		return err
	}

	var decoded any
	if err := json.Unmarshal(encoded, &decoded); err != nil {
		return err
	}

	issues := make([]string, 0)
	validateView(decoded, "view", &issues)
	if len(issues) > 0 {
		return ValidationError{Issues: issues}
	}
	return nil
}

func ValidatePolicy(input any, policy ViewPolicy) error {
	encoded, err := json.Marshal(input)
	if err != nil {
		return err
	}

	var decoded any
	if err := json.Unmarshal(encoded, &decoded); err != nil {
		return err
	}

	if err := Validate(decoded); err != nil {
		return err
	}

	issues := make([]string, 0)
	validateViewPolicy(decoded, "view", policy, &issues)
	if len(issues) > 0 {
		return PolicyError{Issues: issues}
	}
	return nil
}

func validateView(input any, path string, issues *[]string) {
	view, ok := input.(map[string]any)
	if !ok {
		*issues = append(*issues, fmt.Sprintf("%s must be an object", path))
		return
	}

	validateAllowedKeys(view, []string{"template", "state", "actions", "meta"}, path, issues)

	if _, ok := view["template"].(string); !ok {
		*issues = append(*issues, fmt.Sprintf("%s.template must be a string", path))
	}

	state, ok := view["state"].(map[string]any)
	if !ok {
		*issues = append(*issues, fmt.Sprintf("%s.state must be an object", path))
	} else {
		validateJSONObject(state, path+".state", issues)
	}

	if actionsValue, exists := view["actions"]; exists {
		actions, ok := actionsValue.(map[string]any)
		if !ok {
			*issues = append(*issues, fmt.Sprintf("%s.actions must be an object", path))
		} else {
			for actionName, descriptor := range actions {
				validateAction(descriptor, path+".actions."+actionName, issues)
			}
		}
	}

	if metaValue, exists := view["meta"]; exists {
		validateMeta(metaValue, path+".meta", issues)
	}
}

func validateAction(input any, path string, issues *[]string) {
	action, ok := input.(map[string]any)
	if !ok {
		*issues = append(*issues, fmt.Sprintf("%s must be an object", path))
		return
	}

	actionType, ok := action["type"].(string)
	if !ok {
		*issues = append(*issues, fmt.Sprintf("%s.type must be a string", path))
		return
	}

	switch actionType {
	case "fetch":
		validateAllowedKeys(action, []string{"type", "url", "method", "payload"}, path, issues)
		if _, ok := action["url"].(string); !ok {
			*issues = append(*issues, fmt.Sprintf("%s.url must be a string", path))
		}
		if method, exists := action["method"]; exists {
			methodString, ok := method.(string)
			if !ok || !contains([]string{"GET", "POST", "PUT", "PATCH", "DELETE"}, methodString) {
				*issues = append(*issues, fmt.Sprintf("%s.method must be a supported HTTP method", path))
			}
		}
		if payload, exists := action["payload"]; exists {
			payloadMap, ok := payload.(map[string]any)
			if !ok {
				*issues = append(*issues, fmt.Sprintf("%s.payload must be an object", path))
			} else {
				validateJSONObject(payloadMap, path+".payload", issues)
			}
		}
	case "emit":
		validateAllowedKeys(action, []string{"type", "event", "detail"}, path, issues)
		if _, ok := action["event"].(string); !ok {
			*issues = append(*issues, fmt.Sprintf("%s.event must be a string", path))
		}
		if detail, exists := action["detail"]; exists {
			detailMap, ok := detail.(map[string]any)
			if !ok {
				*issues = append(*issues, fmt.Sprintf("%s.detail must be an object", path))
			} else {
				validateJSONObject(detailMap, path+".detail", issues)
			}
		}
	case "navigate":
		validateAllowedKeys(action, []string{"type", "url", "replace"}, path, issues)
		if _, ok := action["url"].(string); !ok {
			*issues = append(*issues, fmt.Sprintf("%s.url must be a string", path))
		}
		if replace, exists := action["replace"]; exists {
			if _, ok := replace.(bool); !ok {
				*issues = append(*issues, fmt.Sprintf("%s.replace must be a boolean", path))
			}
		}
	default:
		*issues = append(*issues, fmt.Sprintf("%s.type must be one of fetch, emit, navigate", path))
	}
}

func validateMeta(input any, path string, issues *[]string) {
	metaValue, ok := input.(map[string]any)
	if !ok {
		*issues = append(*issues, fmt.Sprintf("%s must be an object", path))
		return
	}

	validateAllowedKeys(metaValue, []string{"version", "lang", "generator", "title", "route"}, path, issues)

	if version, exists := metaValue["version"]; exists {
		versionString, ok := version.(string)
		if !ok || versionString != "1" {
			*issues = append(*issues, fmt.Sprintf("%s.version must be \"1\"", path))
		}
	}

	for _, key := range []string{"lang", "generator", "title", "route"} {
		if value, exists := metaValue[key]; exists {
			if _, ok := value.(string); !ok {
				*issues = append(*issues, fmt.Sprintf("%s.%s must be a string", path, key))
			}
		}
	}
}

func validateJSONObject(input map[string]any, path string, issues *[]string) {
	for key, value := range input {
		validateJSONValue(value, path+"."+key, issues)
	}
}

func validateJSONValue(value any, path string, issues *[]string) {
	switch typed := value.(type) {
	case nil, string, float64, bool:
		return
	case []any:
		for index, item := range typed {
			validateJSONValue(item, fmt.Sprintf("%s[%d]", path, index), issues)
		}
	case map[string]any:
		for key, item := range typed {
			validateJSONValue(item, path+"."+key, issues)
		}
	default:
		*issues = append(*issues, fmt.Sprintf("%s must be JSON-serializable", path))
	}
}

func validateAllowedKeys(input map[string]any, allowed []string, path string, issues *[]string) {
	for key := range input {
		if !contains(allowed, key) {
			*issues = append(*issues, fmt.Sprintf("%s contains unsupported property %q", path, key))
		}
	}
}

func validateViewPolicy(input any, path string, policy ViewPolicy, issues *[]string) {
	view, ok := input.(map[string]any)
	if !ok {
		return
	}

	template, _ := view["template"].(string)
	meta, _ := view["meta"].(map[string]any)

	if policy.RequireVersion {
		if version, ok := meta["version"].(string); !ok || version != "1" {
			*issues = append(*issues, fmt.Sprintf("%s.meta.version is required by policy", path))
		}
	}
	if policy.RequireTitle {
		if title, ok := meta["title"].(string); !ok || title == "" {
			*issues = append(*issues, fmt.Sprintf("%s.meta.title is required by policy", path))
		}
	}
	if policy.RequireRoute {
		if route, ok := meta["route"].(string); !ok || route == "" {
			*issues = append(*issues, fmt.Sprintf("%s.meta.route is required by policy", path))
		}
	}
	if !policy.AllowSafeFilter && safeFilterPattern.MatchString(template) {
		*issues = append(*issues, fmt.Sprintf("%s.template uses the |safe filter, which is disabled by policy", path))
	}
	if !policy.AllowScriptTags && scriptTagPattern.MatchString(template) {
		*issues = append(*issues, fmt.Sprintf("%s.template contains a <script> tag, which is disabled by policy", path))
	}
	if !policy.AllowInlineHandlers && inlineHandlerPattern.MatchString(template) {
		*issues = append(*issues, fmt.Sprintf("%s.template contains inline event handler attributes, which are disabled by policy", path))
	}
	if !policy.AllowJavaScriptURLs && javascriptURLPattern.MatchString(template) {
		*issues = append(*issues, fmt.Sprintf("%s.template contains a javascript: URL, which is disabled by policy", path))
	}

	if policy.ActionURLs != ActionURLsRelativeOnly {
		return
	}

	actions, ok := view["actions"].(map[string]any)
	if !ok {
		return
	}
	for actionName, raw := range actions {
		validateActionPolicy(raw, path+".actions."+actionName, issues)
	}
}

func validateActionPolicy(input any, path string, issues *[]string) {
	action, ok := input.(map[string]any)
	if !ok {
		return
	}

	actionType, _ := action["type"].(string)
	if actionType != "fetch" && actionType != "navigate" {
		return
	}

	url, ok := action["url"].(string)
	if !ok || !isRelativeURL(url) {
		*issues = append(*issues, fmt.Sprintf("%s.url must stay relative under the current policy", path))
	}
}

func contains(values []string, target string) bool {
	for _, value := range values {
		if value == target {
			return true
		}
	}
	return false
}

func isRelativeURL(url string) bool {
	if strings.HasPrefix(url, "//") {
		return false
	}
	return strings.HasPrefix(url, "/") ||
		strings.HasPrefix(url, "./") ||
		strings.HasPrefix(url, "../") ||
		strings.HasPrefix(url, "?") ||
		strings.HasPrefix(url, "#")
}
